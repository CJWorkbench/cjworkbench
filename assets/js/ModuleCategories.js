/**
* Returns an array of <Module Category> components,
*  each of which has child <Module>, sorted by type.
*
* Currently rendered by <ModuleLibraryClosed> and <ModuleLibraryClosed> components
*
*/

import PropTypes from 'prop-types'
import React from 'react'
import ModuleCategory from './ModuleCategory'
import Module from './Module'

function modulesToCategories(modules) {
  const ret = []
  const keyed = {}

  for (const module of modules) {
    const categoryName = module.category

    let category
    if (keyed.hasOwnProperty(categoryName)) {
      category = keyed[categoryName]
    } else {
      category = keyed[categoryName] = {
        name: categoryName,
        modules: [],
      }
      ret.push(category)
    }

    category.modules.push(module)
  }

  return ret
}

export default class ModuleCategories extends React.Component {
  _renderCategory(category) {
    const { name, modules } = category
    const { isReadOnly, libraryOpen, addModule, dropModule, setOpenCategory } = this.props
    const collapsed = this.props.openCategory !== name

    return (
      <ModuleCategory
        name={name}
        key={name}
        modules={modules}
        isReadOnly={isReadOnly}
        collapsed={collapsed}
        addModule={addModule}
        dropModule={dropModule}
        setOpenCategory={setOpenCategory}
        libraryOpen={libraryOpen}
        />
    )
  }

  render() {
    const categories = modulesToCategories(this.props.modules)
    const categoryComponents = categories.map(c => this._renderCategory(c))

    return <div className="list">{categoryComponents}</div>
  }
}

ModuleCategories.propTypes = {
  openCategory:     PropTypes.string,
  addModule:        PropTypes.func.isRequired,
  dropModule:       PropTypes.func.isRequired,
  modules:          PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired,
    icon: PropTypes.string.isRequired,
  })).isRequired,
  setOpenCategory:  PropTypes.func.isRequired,
  libraryOpen:      PropTypes.bool.isRequired,
  isReadOnly:       PropTypes.bool.isRequired,
}
