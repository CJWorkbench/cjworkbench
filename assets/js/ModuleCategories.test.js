/**
 * Testing Stories:
 * -Renders list of <ModuleCategory> components
 * 
 */

import React from 'react'
import ModuleCategories  from './ModuleCategories'
import { mount } from 'enzyme'
import HTML5Backend from 'react-dnd-html5-backend'
import { DragDropContextProvider } from 'react-dnd'


describe('ModuleCategories ', () => {
  
  var wrapper;  
  var items = [
    {
      "id":4,
      "name":"Load from Enigma",
      "category":"Add data",
      "description":"Connect a dataset from Enigma's collection via URL.",
      "link":"",
      "author":"Workbench",
      "icon":"url"
    },
    {
      "id":10,
      "name":"Filter by Text",
      "category":"Filter",
      "description":"Filter rows by matching text in specific columns.",
      "link":"",
      "author":"Workbench",
      "icon":"filter"
    }
  ];   
  
  describe('Library open ', () => {

    beforeEach(() => wrapper = mount(
      <DragDropContextProvider backend={HTML5Backend}>
        <ModuleCategories
          openCategory={"Add Data"} 
          setOpenCategory={() => {}}
          libraryOpen={true}
          isReadOnly={false}            
          addModule={() => {}}
          dropModule={() => {}}
          items={items}
        />
      </DragDropContextProvider>
    ));

    afterEach(() => wrapper.unmount());    
  
    it('Renders with list of ModuleCategory components', () => { 
      expect(wrapper).toMatchSnapshot();
    });

  });

  describe('Library closed ', () => {
  
    beforeEach(() => wrapper = mount(
      <DragDropContextProvider backend={HTML5Backend}>
        <ModuleCategories
          openCategory={"Add Data"} 
          setOpenCategory={() => {}}
          libraryOpen={false}
          isReadOnly={false}            
          addModule={() => {}}
          dropModule={() => {}}
          items={items}
        />
      </DragDropContextProvider>
    ));

    afterEach(() => wrapper.unmount());    
  
    it('Renders with list of ModuleCategory components', () => { 
      expect(wrapper).toMatchSnapshot();
    });
  
  });
      
});
