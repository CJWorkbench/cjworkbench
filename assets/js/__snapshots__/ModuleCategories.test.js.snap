// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleCategories  Library closed  Renders with list of ModuleCategory components 1`] = `
<DragDropContextProvider
  backend={[Function]}
>
  <ModuleCategories
    addModule={[Function]}
    dropModule={[Function]}
    isReadOnly={false}
    items={
      Array [
        Object {
          "author": "Workbench",
          "category": "Add data",
          "description": "Connect a dataset from Enigma's collection via URL.",
          "icon": "url",
          "id": 4,
          "link": "",
          "name": "Load from Enigma",
        },
        Object {
          "author": "Workbench",
          "category": "Filter",
          "description": "Filter rows by matching text in specific columns.",
          "icon": "filter",
          "id": 10,
          "link": "",
          "name": "Filter by Text",
        },
      ]
    }
    libraryOpen={false}
    openCategory="Add Data"
    setOpenCategory={[Function]}
  >
    <div
      className="list"
    >
      <ModuleCategory
        collapsed={true}
        isReadOnly={false}
        libraryOpen={false}
        modules={
          Array [
            <DragSource(Module)
              addModule={[Function]}
              dropModule={[Function]}
              icon="url"
              id={4}
              name="Load from Enigma"
          />,
          ]
        }
        name="Add data"
        setOpenCategory={[Function]}
      >
        <div
          className="card b-l-gray library-card-category-closed"
        >
          <div
            className="ml-cat"
          >
            <div
              className="first-level"
              onMouseEnter={[Function]}
            >
              <div
                className="closed-ML-cat"
              >
                <span
                  className="icon-url ml-icon"
                />
              </div>
            </div>
            <div
              className="ml-list-mini"
              style={
                Object {
                  "display": "none",
                }
              }
            >
              <DragSource(Module)
                addModule={[Function]}
                dropModule={[Function]}
                icon="url"
                id={4}
                name="Load from Enigma"
              >
                <Module
                  addModule={[Function]}
                  connectDragSource={[Function]}
                  dropModule={[Function]}
                  icon="url"
                  id={4}
                  isDragging={false}
                  name="Load from Enigma"
                >
                  <div
                    className="card ml-module-card"
                  >
                    <div
                      className=""
                      onClick={[Function]}
                    >
                      <div
                        className="second-level d-flex"
                      >
                        <div
                          className="d-flex flex-row align-items-center"
                        >
                          <div
                            className="ml-icon-container"
                          >
                            <div
                              className="icon-url ml-icon"
                            />
                          </div>
                          <div>
                            <div
                              className="content-5 ml-module-name"
                            >
                              Load from Enigma
                            </div>
                          </div>
                        </div>
                        <div
                          className="ml-handle"
                        >
                          <div
                            className="icon-grip"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </Module>
              </DragSource(Module)>
            </div>
          </div>
        </div>
      </ModuleCategory>
      <ModuleCategory
        collapsed={true}
        isReadOnly={false}
        libraryOpen={false}
        modules={
          Array [
            <DragSource(Module)
              addModule={[Function]}
              dropModule={[Function]}
              icon="filter"
              id={10}
              name="Filter by Text"
          />,
          ]
        }
        name="Filter"
        setOpenCategory={[Function]}
      >
        <div
          className="card b-l-gray library-card-category-closed"
        >
          <div
            className="ml-cat"
          >
            <div
              className="first-level"
              onMouseEnter={[Function]}
            >
              <div
                className="closed-ML-cat"
              >
                <span
                  className="icon-filter ml-icon"
                />
              </div>
            </div>
            <div
              className="ml-list-mini"
              style={
                Object {
                  "display": "none",
                }
              }
            >
              <DragSource(Module)
                addModule={[Function]}
                dropModule={[Function]}
                icon="filter"
                id={10}
                name="Filter by Text"
              >
                <Module
                  addModule={[Function]}
                  connectDragSource={[Function]}
                  dropModule={[Function]}
                  icon="filter"
                  id={10}
                  isDragging={false}
                  name="Filter by Text"
                >
                  <div
                    className="card ml-module-card"
                  >
                    <div
                      className=""
                      onClick={[Function]}
                    >
                      <div
                        className="second-level d-flex"
                      >
                        <div
                          className="d-flex flex-row align-items-center"
                        >
                          <div
                            className="ml-icon-container"
                          >
                            <div
                              className="icon-filter ml-icon"
                            />
                          </div>
                          <div>
                            <div
                              className="content-5 ml-module-name"
                            >
                              Filter by Text
                            </div>
                          </div>
                        </div>
                        <div
                          className="ml-handle"
                        >
                          <div
                            className="icon-grip"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </Module>
              </DragSource(Module)>
            </div>
          </div>
        </div>
      </ModuleCategory>
    </div>
  </ModuleCategories>
</DragDropContextProvider>
`;

exports[`ModuleCategories  Library open  Renders with list of ModuleCategory components 1`] = `
<DragDropContextProvider
  backend={[Function]}
>
  <ModuleCategories
    addModule={[Function]}
    dropModule={[Function]}
    isReadOnly={false}
    items={
      Array [
        Object {
          "author": "Workbench",
          "category": "Add data",
          "description": "Connect a dataset from Enigma's collection via URL.",
          "icon": "url",
          "id": 4,
          "link": "",
          "name": "Load from Enigma",
        },
        Object {
          "author": "Workbench",
          "category": "Filter",
          "description": "Filter rows by matching text in specific columns.",
          "icon": "filter",
          "id": 10,
          "link": "",
          "name": "Filter by Text",
        },
      ]
    }
    libraryOpen={true}
    openCategory="Add Data"
    setOpenCategory={[Function]}
  >
    <div
      className="list"
    >
      <ModuleCategory
        collapsed={true}
        isReadOnly={false}
        libraryOpen={true}
        modules={
          Array [
            <DragSource(Module)
              addModule={[Function]}
              dropModule={[Function]}
              icon="url"
              id={4}
              name="Load from Enigma"
          />,
          ]
        }
        name="Add data"
        setOpenCategory={[Function]}
      >
        <div
          className="card b-l-gray library-card-category-closed"
        >
          <div
            className="ml-cat"
          >
            <div
              className="first-level"
              onClick={[Function]}
            >
              <div
                className="cat-container"
              >
                <span
                  className="open-ML-cat"
                >
                  Add data
                </span>
                <div
                  className="ml-sort icon-sort-down-vl-gray"
                />
              </div>
            </div>
            <Collapse
              delay={
                Object {
                  "hide": 350,
                  "show": 350,
                }
              }
              isOpen={false}
              onClosed={[Function]}
              onOpened={[Function]}
              tag="div"
            >
              <div
                className="collapse"
                style={Object {}}
              >
                <div
                  className="ml-list"
                >
                  <DragSource(Module)
                    addModule={[Function]}
                    dropModule={[Function]}
                    icon="url"
                    id={4}
                    name="Load from Enigma"
                  >
                    <Module
                      addModule={[Function]}
                      connectDragSource={[Function]}
                      dropModule={[Function]}
                      icon="url"
                      id={4}
                      isDragging={false}
                      name="Load from Enigma"
                    >
                      <div
                        className="card ml-module-card"
                      >
                        <div
                          className=""
                          onClick={[Function]}
                        >
                          <div
                            className="second-level d-flex"
                          >
                            <div
                              className="d-flex flex-row align-items-center"
                            >
                              <div
                                className="ml-icon-container"
                              >
                                <div
                                  className="icon-url ml-icon"
                                />
                              </div>
                              <div>
                                <div
                                  className="content-5 ml-module-name"
                                >
                                  Load from Enigma
                                </div>
                              </div>
                            </div>
                            <div
                              className="ml-handle"
                            >
                              <div
                                className="icon-grip"
                              />
                            </div>
                          </div>
                        </div>
                      </div>
                    </Module>
                  </DragSource(Module)>
                </div>
              </div>
            </Collapse>
          </div>
        </div>
      </ModuleCategory>
      <ModuleCategory
        collapsed={true}
        isReadOnly={false}
        libraryOpen={true}
        modules={
          Array [
            <DragSource(Module)
              addModule={[Function]}
              dropModule={[Function]}
              icon="filter"
              id={10}
              name="Filter by Text"
          />,
          ]
        }
        name="Filter"
        setOpenCategory={[Function]}
      >
        <div
          className="card b-l-gray library-card-category-closed"
        >
          <div
            className="ml-cat"
          >
            <div
              className="first-level"
              onClick={[Function]}
            >
              <div
                className="cat-container"
              >
                <span
                  className="open-ML-cat"
                >
                  Filter
                </span>
                <div
                  className="ml-sort icon-sort-down-vl-gray"
                />
              </div>
            </div>
            <Collapse
              delay={
                Object {
                  "hide": 350,
                  "show": 350,
                }
              }
              isOpen={false}
              onClosed={[Function]}
              onOpened={[Function]}
              tag="div"
            >
              <div
                className="collapse"
                style={Object {}}
              >
                <div
                  className="ml-list"
                >
                  <DragSource(Module)
                    addModule={[Function]}
                    dropModule={[Function]}
                    icon="filter"
                    id={10}
                    name="Filter by Text"
                  >
                    <Module
                      addModule={[Function]}
                      connectDragSource={[Function]}
                      dropModule={[Function]}
                      icon="filter"
                      id={10}
                      isDragging={false}
                      name="Filter by Text"
                    >
                      <div
                        className="card ml-module-card"
                      >
                        <div
                          className=""
                          onClick={[Function]}
                        >
                          <div
                            className="second-level d-flex"
                          >
                            <div
                              className="d-flex flex-row align-items-center"
                            >
                              <div
                                className="ml-icon-container"
                              >
                                <div
                                  className="icon-filter ml-icon"
                                />
                              </div>
                              <div>
                                <div
                                  className="content-5 ml-module-name"
                                >
                                  Filter by Text
                                </div>
                              </div>
                            </div>
                            <div
                              className="ml-handle"
                            >
                              <div
                                className="icon-grip"
                              />
                            </div>
                          </div>
                        </div>
                      </div>
                    </Module>
                  </DragSource(Module)>
                </div>
              </div>
            </Collapse>
          </div>
        </div>
      </ModuleCategory>
    </div>
  </ModuleCategories>
</DragDropContextProvider>
`;
